ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs

```diff
@@ -2,353 +2,395 @@
+		public delegate void OutputDeviceEventSent(MidiEventSentEventArgs e);
+
+		/// <summary>
+		/// The playback selected track notification delegate.
+		/// </summary>
+		/// <param name="e">The event arguments.</param>
+		public delegate void PlaybackSelectedTrackNotification(TrackNotificationEventArgs e);
+
+		/// <summary>
+		/// The render current notes delegate.
+		/// </summary>
+		/// <param name="queue">The queue of markable playback events.</param>
+		/// <param name="currentTime">The current time of playback.</param>
+		/// <returns>An asynchronous task.</returns>
+		public delegate Task RenderCurrentNotesAsync(ConcurrentQueue<MarkablePlaybackEvent>[] queue, TimeSpan currentTime);
+
+		/// <summary>
+		/// The output device sent event.
+		/// </summary>
+		public event OutputDeviceEventSent OutputDeviceEventSentEvent;
+
+		/// <summary>
+		/// The playback selected track notification event.
+		/// </summary>
+		public event PlaybackSelectedTrackNotification PlaybackSelectedTrackNotificationEvent;
+
+		/// <summary>
+		/// The render current notes event.
+		/// </summary>
+		public event RenderCurrentNotesAsync RenderCurrentNotesAsyncEvent;
+
+		/// <summary>
+		/// The playback finished.
+		/// </summary>
+		public event EventHandler<EventArgs> PlaybackFinished;
+
+		/// <summary>
+		/// Gets or sets the MIDI file.
+		/// </summary>
+		public StorageFile File { get; set; }
+
+		/// <summary>
+		/// Gets or sets the display name of first track.
+		/// </summary>
+		public string TrackDisplayName { get; set; }
+
+		/// <summary>
+		/// Gets the selected track/channel.
+		/// </summary>
+		public FourBitNumber SelectedTrack => this.playback.SelectedChannel;
+
+		/// <summary>
+		/// Gets the set of instruments associated with the MIDI file.
+		/// </summary>
+		public IDictionary<FourBitNumber, InstrumentInfo> Instruments { get; } =
+			new Dictionary<FourBitNumber, InstrumentInfo>();
+
+		/// <summary>
+		/// Gets the set of pitches associated with the selected instrument.
+		/// </summary>
+		/// <param name="instrument">Instrument to retrieve pitches for.</param>
+		/// <returns>Pitches for instrument.</returns>
+		public List<string> RetrievePitchesForInstrument(FourBitNumber instrument) =>
+			this.midi.GetNotes()
+				.Where(note => note.Channel == instrument)
+				.Select(note => note.NoteName.ToString())
+				.Distinct()
+				.ToList();
+
+		/// <summary>
+		/// Gets the play times of notes associated with the selected instrument.
+		/// </summary>
+		/// <param name="instrument">Instrument to retrieve note times for.</param>
+		/// <returns>Note times for instrument.</returns>
+		public List<long> RetrieveNoteTimesForInstrument(FourBitNumber instrument)
+		{
+			var midiMap = this.midi.GetTempoMap();
+			return this.midi.GetNotes()
+				.Where(note => note.Channel == instrument)
+				.Select(note =>
+				{
+					var ms = ((MetricTimeSpan)note.TimeAs(TimeSpanType.Metric, midiMap)).TotalMicroseconds / 1000;
```

Comment:
You can just use `.Select(note => ((MetricTimeSpan)note.TimeAs(TimeSpanType.Metric, midiMap)).TotalMicroseconds / 1000)`